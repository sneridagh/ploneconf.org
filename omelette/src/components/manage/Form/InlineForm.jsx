import PropTypes from 'prop-types';
import { defineMessages, injectIntl } from 'react-intl';
import React from 'react';
import { keys, map } from 'lodash';
import { Field } from '@plone/volto/components';
import { Segment, Message } from 'semantic-ui-react';

const messages = defineMessages({
  editValues: {
    id: 'Edit values',
    defaultMessage: 'Edit values',
  },
  error: {
    id: 'Error',
    defaultMessage: 'Error',
  },
  thereWereSomeErrors: {
    id: 'There were some errors',
    defaultMessage: 'There were some errors',
  },
});

const InlineForm = ({
  block,
  description,
  error, // Such as {message: "It's not good"}
  errors = {},
  formData,
  onChangeField,
  schema,
  title,
  icon,
  headerActions,
  footer,
  intl,
}) => {
  const _ = intl.formatMessage;
  const defaultFieldset = schema.fieldsets.find((o) => o.id === 'default');
  const other = schema.fieldsets.filter((o) => o.id !== 'default');
  return (
    <Segment.Group raised className="form">
      <header className="header pulled">
        {icon}
        <h2>{title || _(messages.editValues)}</h2>
        {headerActions}
      </header>
      {description && (
        <Segment secondary className="attached">
          {description}
        </Segment>
      )}
      {keys(errors).length > 0 && (
        <Message
          icon="warning"
          negative
          attached
          header={_(messages.error)}
          content={_(messages.thereWereSomeErrors)}
        />
      )}
      {error && (
        <Message
          icon="warning"
          negative
          attached
          header={_(messages.error)}
          content={error.message}
        />
      )}

      <div id={`blockform-fieldset-${defaultFieldset.id}`}>
        {map(defaultFieldset.fields, (field, index) => (
          <Field
            {...schema.properties[field]}
            id={field}
            fieldSet={defaultFieldset.title.toLowerCase()}
            focus={index === 0}
            value={formData[field] || schema.properties[field].default}
            required={schema.required.indexOf(field) !== -1}
            onChange={(id, value) => {
              onChangeField(id, value);
            }}
            key={field}
            error={errors[field]}
            block={block}
          />
        ))}
      </div>

      {other.map((fieldset) => (
        <div key={fieldset.id} id={`blockform-fieldset-${fieldset.id}`}>
          {title && (
            <Segment className="secondary attached">{fieldset.title}</Segment>
          )}
          <Segment className="attached">
            {map(fieldset.fields, (field) => (
              <Field
                {...schema.properties[field]}
                id={field}
                value={formData[field] || schema.properties[field].default}
                required={schema.required.indexOf(field) !== -1}
                onChange={(id, value) => {
                  onChangeField(id, value);
                }}
                key={field}
                error={errors[field]}
                block={block}
              />
            ))}
          </Segment>
        </div>
      ))}
      {footer}
    </Segment.Group>
  );
};

InlineForm.defaultProps = {
  block: null,
  description: null,
  formData: null,
  onChangeField: null,
  error: null,
  errors: {},
  schema: {},
};

InlineForm.propTypes = {
  block: PropTypes.string,
  description: PropTypes.string,
  schema: PropTypes.shape({
    fieldsets: PropTypes.arrayOf(
      PropTypes.shape({
        fields: PropTypes.arrayOf(PropTypes.string),
        id: PropTypes.string,
        title: PropTypes.string,
      }),
    ),
    properties: PropTypes.objectOf(PropTypes.any),
    definitions: PropTypes.objectOf(PropTypes.any),
    required: PropTypes.arrayOf(PropTypes.string),
  }),
  formData: PropTypes.objectOf(PropTypes.any),
  pathname: PropTypes.string,
  onChangeField: PropTypes.func,
  error: PropTypes.shape({
    message: PropTypes.string,
  }),
};

export default injectIntl(InlineForm, { forwardRef: true });
