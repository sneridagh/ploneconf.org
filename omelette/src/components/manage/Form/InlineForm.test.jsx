import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-intl-redux';

import InlineForm from './InlineForm';

const mockStore = configureStore();

jest.mock('@plone/volto/components/manage/Form/Field', () =>
  jest.fn(() => <div className="Field" />),
);

describe('Form', () => {
  it('renders a form component', () => {
    const store = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
    });
    const component = renderer.create(
      <Provider store={store}>
        <InlineForm
          schema={{
            fieldsets: [
              {
                id: 'default',
                title: 'Default',
                fields: ['title'],
              },
            ],
            properties: {
              title: {},
            },
            required: [],
          }}
          onSubmit={() => {}}
          onCancel={() => {}}
          formData={{}}
        />
      </Provider>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});
