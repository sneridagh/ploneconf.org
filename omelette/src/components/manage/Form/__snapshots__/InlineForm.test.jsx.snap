// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders a form component 1`] = `
<div
  className="ui raised segments form"
>
  <header
    className="header pulled"
  >
    <h2>
      Edit values
    </h2>
  </header>
  <div
    id="blockform-fieldset-default"
  >
    <div
      className="Field"
    />
  </div>
</div>
`;
